source/common/config/watch_map.cc

```diff
@@ -75,12 +84,21 @@ void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>
   for (auto& watch : watches_) {
     const auto this_watch_updates = per_watch_updates.find(watch);
     if (this_watch_updates == per_watch_updates.end()) {
-      // This update included no resources this watch cares about - so we do an empty
-      // onConfigUpdate(), to notify the watch that its resources - if they existed before this -
-      // were dropped.
-      watch->callbacks_.onConfigUpdate({}, version_info);
+      // This update included no resources this watch cares about.
+      // 1) If there is only a single, wildcard watch (i.e. Cluster or Listener), always call
+      //    its onConfigUpdate even if just a no-op, to properly maintain state-of-the-world
+      //    semantics and the update_empty stat.
+      // 2) If this watch previously had some resources, it means this update is removing all
+      //    of this watch's resources, so the watch must be informed with an onConfigUpdate.
+      // 3) Otherwise, we can skip onConfigUpdate for this watch.
+      if ((watches_.size() == 1 && wildcard_watches_.size() == 1) ||
```

Comment:
Better assign a name for `(watches_.size() == 1 && wildcard_watches_.size() == 1)` outside the for loop. The condition is not bind to a single loop, but the state of all watches.

Reply:
Thanks, excellent suggestion, I like it much more this way.